src/VisualStudio/Core/Def/Implementation/ProjectSystem/MiscellaneousFilesWorkspace.cs

```diff
@@ -207,79 +188,155 @@ public int OnBeforeLastDocumentUnlock(uint docCookie, uint dwRDTLockType, uint d
+            if (workspaceRegistration.Workspace == null)
+            {
+                // We should now claim this
+                AttachToDocument(docCookie, moniker);
+            }
+            else if (IsClaimedByAnotherWorkspace(workspaceRegistration))
+            {
+                // It's now claimed by another workspace, so we should unclaim it
+                if (_docCookiesToHostProject.ContainsKey(docCookie))
                 {
-                    parseOptions = parseOptions.WithKind(SourceCodeKind.Script);
+                    DetachFromDocument(docCookie, moniker);
                 }
+            }
+        }
 
-                // First, create the project
-                var hostProject = new HostProject(this, CurrentSolution.Id, languageInformation.LanguageName, parseOptions, _metadataReferences);
+        /// <summary>
+        /// Stops tracking a document in the RDT for whether we should attach to it.
+        /// </summary>
+        /// <returns>true if we were previously tracking it or not.</returns>
+        private bool TryUntrackClosingDocument(uint docCookie, string moniker)
+        {
+            bool unregisteredRegistration = false;
 
-                // Now try to find the document. We accept any text buffer, since we've already verified it's an appropriate file in ShouldIncludeFile.
-                var document = _documentProvider.TryGetDocumentForFile(hostProject, (uint)VSConstants.VSITEMID.Nil, moniker, parseOptions.Kind, t => true);
+            // Remove our registration changing before we call DetachFromDocument. Otherwise, calling DetachFromDocument
+            // causes us to set the workspace to null, which we then respond to as an indication that we should
+            // attach again.
+            WorkspaceRegistration registration;
+            if (_docCookieToWorkspaceRegistration.TryGetValue(docCookie, out registration))
+            {
+                registration.WorkspaceChanged -= Registration_WorkspaceChanged;
+                _docCookieToWorkspaceRegistration = _docCookieToWorkspaceRegistration.RemoveKey(docCookie);
+                unregisteredRegistration = true;
+            }
 
-                // If the buffer has not yet been initialized, we won't get a document.
-                if (document == null)
-                {
-                    return;
-                }
+            DetachFromDocument(docCookie, moniker);
+
+            return unregisteredRegistration;
+        }
+
+        private bool IsClaimedByAnotherWorkspace(WorkspaceRegistration registration)
+        {
+            // Currently, we are also responsible for pushing documents to the metadata as source workspace,
+            // so we count that here as well
+            return registration.Workspace != null && registration.Workspace.Kind != WorkspaceKind.MetadataAsSource && registration.Workspace.Kind != WorkspaceKind.MiscellaneousFiles;
```

Comment:
Can you change that to `&& registration.Workspace != this`?

It might be useful for custom workspaces to claim to be `WorkspaceKind.MiscellaneousFiles` (I don't remember offhand what services each one claims, but there are scenarios where a custom workspace will want a subset of the VS host services)